### PR TITLE
[DOC] Update plot_localizer_mass_univariate_methods.py

### PR DESCRIPTION
**Pull request by emdupre**
- Partially addresses #6265 

Use of AI tools:
- [ ] LLM tools were used to generate at least part of the content of this pull-request.
- [X] No LLM tools were used to generate any part of the content of this pull-request.

Changes proposed in this pull request:

- Updates _Massively univariate analysis of a motor task from the Localizer dataset_ example
- Adds more links to glossary, scikit-learn, and wikipedia as necessary
- Adds more explanatory text

### CODE REVIEW

File: examples/07_advanced/plot_localizer_mass_univariate_methods.py [hunk 1]
```diff
@@ -2,69 +2,122 @@
...
-   Data smoothed at 5 :term:`voxels<voxel>` :term:`FWHM` are used.
+This example compares results obtained with a massively univariate
+analysis (:func:`~nilearn.mass_univariate.permuted_ols`)
+after two permutation test correction methods:
+Max :term:`t-statistic <Parameter estimate>` Family-wise Error
+(:term:`FWE <FWER correction>`) and
+Max :term:`TFCE` :term:`FWE <FWER correction>`.
+
+These two methods are compared against a baseline parametric test
+(Bonferroni-Holm :term:`FWE <FWER correction>`).
```
Comment (emdupre) [inline examples/07_advanced/plot_localizer_mass_univariate_methods.py]: @bthirion the original text said "Bonferroni FWE", which I think would be Bonferroni-Holm. Is this actually a Bonferroni-Holm correction, though ?

File: examples/07_advanced/plot_localizer_mass_univariate_methods.py [hunk 2]
```diff
@@ -2,69 +2,122 @@
...
+# :func:`~nilearn.maskers.MultiNiftiMasker` object in order to
+# extract all :term:`voxel` values for all subjects into
+# the same array.
+#
+from nilearn.maskers import NiftiMasker
+
 nifti_masker = NiftiMasker(
-    smoothing_fwhm=5, memory="nilearn_cache", memory_level=1, verbose=1
+    smoothing_fwhm=5,
+    standardize="zscore_sample",
```
Comment (emdupre) [inline examples/07_advanced/plot_localizer_mass_univariate_methods.py]: @Remi-Gau I don't see this example listed in #6492, but I added it in to be safe. Should I revert this change ?

**Comment by emdupre** [comment]
👉  [The built example is visible here](https://output.circle-artifacts.com/output/job/4692614e-9848-4f28-90dc-4b08dfca495f/artifacts/0/dev/auto_examples/07_advanced/plot_localizer_mass_univariate_methods.html)

**Comment by emdupre** [comment]
👉 [The built example is visible here](https://output.circle-artifacts.com/output/job/ccee591c-8630-4818-b49b-12d6a4f108c6/artifacts/0/dev/auto_examples/07_advanced/plot_localizer_mass_univariate_methods.html)

File: examples/07_advanced/plot_localizer_mass_univariate_methods.py [hunk 3]
```diff
@@ -2,69 +2,124 @@
...
 )
 fmri_masked = nifti_masker.fit_transform(contrast_map_filenames)
 
 
 # %%
-# Anova (parametric F-scores)
+# :term:`ANOVA` (parametric F-scores)
+# -----------------------------------
+# We use :func:`sklearn.feature_selection.f_regression` to perform
+# :sklearn:`feature selection <modules/modules/feature_selection.html>`
```
Comment (emdupre) [inline examples/07_advanced/plot_localizer_mass_univariate_methods.py]:
```suggestion
# :sklearn:`feature selection <modules/feature_selection.html>`
```

**Comment by emdupre** [comment]
👉  [The built example is visible here](https://output.circle-artifacts.com/output/job/f1bc4810-d05a-4682-86d1-3adb9054097f/artifacts/0/dev/auto_examples/07_advanced/plot_localizer_mass_univariate_methods.html)

File: examples/07_advanced/plot_localizer_mass_univariate_methods.py (hunk 1 above)
Comment (bthirion) [reply examples/07_advanced/plot_localizer_mass_univariate_methods.py]: Actually it's Bonferroni, not Bonferroni-Holm. (actually, we should use Bonferroni-Holm, as it's uniformly more powerful).

**Review by bthirion — COMMENTED** [review]
Just need to revert Bonferroni-Holm to Bonferroni.

File: examples/07_advanced/plot_localizer_mass_univariate_methods.py [hunk 4]
```diff
@@ -2,69 +2,124 @@
...
-    f"at: {localizer_dataset.cmaps[0]}"
-)
+# %%
+behav_var = localizer_dataset.ext_vars["pseudo"].values
+
+# Examine the behavioral variate
+print(behav_var)
 
-tested_var = localizer_dataset.ext_vars["pseudo"]
+# %%%
```
Comment (Remi-Gau) [inline examples/07_advanced/plot_localizer_mass_univariate_methods.py]:
```suggestion
# %%
```

File: examples/07_advanced/plot_localizer_mass_univariate_methods.py [hunk 5]
```diff
@@ -2,69 +2,124 @@
...
+# which yields voxelwise p-values that can be used to assert
+# significance.
+#
 from sklearn.feature_selection import f_regression
 
 _, pvals_anova = f_regression(fmri_masked, tested_var.ravel(), center=True)
+
+# Calculate the negative log of the p-values
+# for equivalent visualization with
+# :func:`~nilearn.mass_univariate.permuted_ols`.
```
Comment (Remi-Gau) [inline examples/07_advanced/plot_localizer_mass_univariate_methods.py:125]:
```suggestion

# %%
# Calculate the negative log of the p-values
# for equivalent visualization with
# :func:`~nilearn.mass_univariate.permuted_ols`.
```

otherwise the ref to the function will not render in HTML

File: examples/07_advanced/plot_localizer_mass_univariate_methods.py [hunk 6]
```diff
@@ -95,22 +154,34 @@
...
+
+# Select first regressor from max :term:`t-statistic <Parameter estimate>`.
 neg_log_pvals_permuted_ols_unmasked = nifti_masker.inverse_transform(
-    ols_outputs["logp_max_t"][0, :]  # select first regressor
+    ols_outputs["logp_max_t"][0, :]
 )
+# Select first regressor from :term:`TFCE`.
 neg_log_pvals_tfce_unmasked = nifti_masker.inverse_transform(
-    ols_outputs["logp_max_tfce"][0, :]  # select first regressor
+    ols_outputs["logp_max_tfce"][0, :]
```
Comment (Remi-Gau) [inline examples/07_advanced/plot_localizer_mass_univariate_methods.py:172]:
```suggestion
# %%
# Select first regressor from max :term:`t-statistic <Parameter estimate>`
# and the first regressor from :term:`TFCE`.
neg_log_pvals_permuted_ols_unmasked = nifti_masker.inverse_transform(
    ols_outputs["logp_max_t"][0, :]
)
neg_log_pvals_tfce_unmasked = nifti_masker.inverse_transform(
    ols_outputs["logp_max_tfce"][0, :]
```

Add a cell for proper rendering of the link to the glossary

File: examples/07_advanced/plot_localizer_mass_univariate_methods.py [hunk 7]
```diff
@@ -95,22 +154,34 @@
...
 )
 
 # %%
 # Visualization
```
Comment (Remi-Gau) [inline examples/07_advanced/plot_localizer_mass_univariate_methods.py:176]: Minor detail: do you think we can adjust the figure to avoid overlap between the colorbar labels and the titles of each subplot

File: examples/07_advanced/plot_localizer_mass_univariate_methods.py [hunk 8]
```diff
@@ -2,69 +2,124 @@
...
-]
-tested_var = tested_var[mask_quality_check].to_numpy().reshape((-1, 1))
+quality_mask = np.isfinite(behav_var)
+n_samples = np.sum(quality_mask)
 print(f"Actual number of subjects after quality check: {int(n_samples)}")
 
+# Cast list of ``cmaps`` to numpy array for Boolean masking.
+contrast_map_filenames = np.array(localizer_dataset.cmaps)[quality_mask]
+# Reshape from shape (n_samples,) to shape (n_samples, 1)
+tested_var = behav_var[quality_mask].reshape((-1, 1))
```
Comment (Remi-Gau) [inline examples/07_advanced/plot_localizer_mass_univariate_methods.py]:
```suggestion
# %%
# Cast list of ``cmaps`` to numpy array for Boolean masking
# and reshape from shape (n_samples,) to shape (n_samples, 1)
contrast_map_filenames = np.array(localizer_dataset.cmaps)[quality_mask]
tested_var = behav_var[quality_mask].reshape((-1, 1))
```

for rendering of ``cmaps``

File: examples/07_advanced/plot_localizer_mass_univariate_methods.py (hunk 2 above)
Comment (Remi-Gau) [reply examples/07_advanced/plot_localizer_mass_univariate_methods.py]: We can clearly see changes in the number of voxels detected with TFCE:

0.13.1:



0.14.0 (stable):



main (dev):



this PR:



But from a quick check on main this seems more due to the fact we have not set random_state for perform_ols. If I do then the number of detection is the same with or without standardization.

Can you confirm you get the same thing in the PR? if so maybe let's not standardize but set the random_state for perform_ols? WDYT?

Comment (Remi-Gau) [reply examples/07_advanced/plot_localizer_mass_univariate_methods.py]: OK I am realizing that none of our perform_ols example set the random_state which also give tiny differences (less noticeable than in here because of the number of permutations is larger than in this one)

See for example plot_oasis_vbm

(check legend for the number of detection)

0.13.1



main - dev 

 

0.14.0 - stable



I am going to open a PR to 'fix' this.

Comment (emdupre) [reply examples/07_advanced/plot_localizer_mass_univariate_methods.py]: Yes, I can confirm that even across multiple local runs we see different results since we have not set a state.
The "standardize" would be more conceptual ; I'm leaning towards reverting it, unless you feel otherwise.

Comment (Remi-Gau) [reply examples/07_advanced/plot_localizer_mass_univariate_methods.py]: let's revert

File: examples/07_advanced/plot_localizer_mass_univariate_methods.py (hunk 7 above)
Comment (emdupre) [reply examples/07_advanced/plot_localizer_mass_univariate_methods.py:176]: I've done a pass to improve this in [f707bd4](https://github.com/nilearn/nilearn/pull/6509/commits/f707bd4282dbf7fb852eaf3b156586cfd0f0db55) ; further suggestions would be welcome !

**Comment by emdupre** [comment]
It looks like this one didn't trigger the CircleCI workflow, but the built artifacts are accessible here : https://github.com/nilearn/nilearn/actions/runs/32741751895/job/97480902769?pr=6509

**Comment by Remi-Gau** [comment]
> It looks like this one didn't trigger the CircleCI workflow, but the built artifacts are accessible here : https://github.com/nilearn/nilearn/actions/runs/32741751895/job/97480902769?pr=6509

sometimes it's a bit slow to appear: now I can see it

**Review by Remi-Gau — APPROVED** [review]
that looks good to me

File: examples/07_advanced/plot_localizer_mass_univariate_methods.py [hunk 9]
```diff
@@ -75,15 +133,19 @@
...
+# :term:`TFCE`-based :term:`FWE <FWER correction>`-corrected
+# negative-log p-values.
 #
 # .. note::
 #   :func:`~nilearn.mass_univariate.permuted_ols` can support a wide range
-#   of analysis designs, depending on the ``tested_var``.
+#   of analysis designs, depending on the numerical labels in ``tested_var``.
 #   For example, if you wished to perform a one-sample test, you could
 #   simply provide an array of ones (e.g., ``np.ones(n_samples)``).
+#
```
Comment (Remi-Gau) [inline examples/07_advanced/plot_localizer_mass_univariate_methods.py:147]: Forgot to mention but this kind of information also has its place in the docstring of permuted_ols, no?

WDYT @emdupre ?

Comment (emdupre) [reply examples/07_advanced/plot_localizer_mass_univariate_methods.py:147]: Ah I thought we updated that in #6501 ! Yes, I think that should be in the docstring, too.
